File: backend/app/schemas/redemption.py
"""兑换码模块 Pydantic schemas（请求/响应）。"""
from __future__ import annotations
from datetime import datetime
from decimal import Decimal
from typing import List, Optional
from pydantic import BaseModel, Field


# ========== 用户端 ==========

class PartnerPublic(BaseModel):
    id: int
    name: str
    description: str
    website_url: str
    logo_url: Optional[str]


class BatchListItem(BaseModel):
    id: int
    partner: PartnerPublic
    name: str
    unit_price: Decimal
    available_count: int


class BatchDetailResponse(BatchListItem):
    description: str


class PurchaseRequest(BaseModel):
    batch_id: int


class PurchaseResponse(BaseModel):
    code_id: int
    code_string: str
    batch_name: str
    partner_name: str
    partner_website_url: str
    description: str
    paid_amount: Decimal
    cash_after: Decimal


class MyRedemptionItem(BaseModel):
    code_id: int
    batch_name: str
    partner_name: str
    partner_website_url: str
    paid_amount: Decimal
    bought_at: datetime
    marked_used_by_user_at: Optional[datetime]


class MyRedemptionDetail(MyRedemptionItem):
    code_string: str
    description: str


class MarkUsedRequest(BaseModel):
    used: bool  # True=标记已用；False=取消标记


# ========== 管理员端 ==========

class PartnerCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    description: str = ""
    website_url: str = ""
    logo_url: Optional[str] = None
    is_active: bool = True


class PartnerUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    description: Optional[str] = None
    website_url: Optional[str] = None
    logo_url: Optional[str] = None
    is_active: Optional[bool] = None


class PartnerAdminItem(BaseModel):
    id: int
    name: str
    description: str
    website_url: str
    logo_url: Optional[str]
    is_active: bool
    created_at: datetime


class BatchCreate(BaseModel):
    partner_id: int
    name: str = Field(min_length=1, max_length=200)
    description: str = ""
    unit_price: Decimal = Field(gt=Decimal("0"))


class BatchUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=200)
    description: Optional[str] = None
    unit_price: Optional[Decimal] = Field(default=None, gt=Decimal("0"))
    status: Optional[str] = None  # draft/active/archived


class BatchAdminItem(BaseModel):
    id: int
    partner_id: int
    partner_name: str
    name: str
    description: str
    unit_price: Decimal
    status: str
    total_count: int
    sold_count: int
    available_count: int
    created_at: datetime


class CsvImportRequest(BaseModel):
    csv_text: str  # 整个 CSV 文本；前端可来自文件读取或粘贴


class CsvImportPreview(BaseModel):
    total_lines: int
    new_codes: List[str]          # 即将插入
    duplicate_codes: List[str]    # 与已有重复（跳过）
    invalid_codes: List[str]      # 非法（空/超长）


class CsvImportConfirm(BaseModel):
    csv_text: str
    confirm: bool = True  # 必须 True 才真正写入


class CsvImportResult(BaseModel):
    inserted: int
    skipped_duplicate: int
    skipped_invalid: int

File: thccb-frontend/src/types/redemption.ts
export interface PartnerPublic {
  id: number
  name: string
  description: string
  website_url: string
  logo_url: string | null
}

export interface BatchListItem {
  id: number
  partner: PartnerPublic
  name: string
  unit_price: string
  available_count: number
}

export interface BatchDetail extends BatchListItem {
  description: string
}

export interface PurchaseResponse {
  code_id: number
  code_string: string
  batch_name: string
  partner_name: string
  partner_website_url: string
  description: string
  paid_amount: string
  cash_after: string
}

export interface MyRedemptionItem {
  code_id: number
  batch_name: string
  partner_name: string
  partner_website_url: string
  paid_amount: string
  bought_at: string
  marked_used_by_user_at: string | null
}

export interface MyRedemptionDetail extends MyRedemptionItem {
  code_string: string
  description: string
}

// ===== Admin =====

export interface PartnerAdminItem {
  id: number
  name: string
  description: string
  website_url: string
  logo_url: string | null
  is_active: boolean
  created_at: string
}

export type BatchStatus = 'draft' | 'active' | 'archived'

export interface BatchAdminItem {
  id: number
  partner_id: number
  partner_name: string
  name: string
  description: string
  unit_price: string
  status: BatchStatus
  total_count: number
  sold_count: number
  available_count: number
  created_at: string
}

export interface CsvImportPreview {
  total_lines: number
  new_codes: string[]
  duplicate_codes: string[]
  invalid_codes: string[]
}

export interface CsvImportResult {
  inserted: number
  skipped_duplicate: number
  skipped_invalid: number
}
